### PR TITLE
Add version 2

### PR DESCRIPTION
**Pull request by OR13**
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-jwt/pull/44.html" title="Last updated on Feb 9, 2023, 1:23 PM UTC (33b5616)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-jwt/44/9bfe4e0...33b5616.html" title="Last updated on Feb 9, 2023, 1:23 PM UTC (33b5616)">Diff</a>

### CODE REVIEW

**Comment by OR13** [comment]
@iherman is there something I need to change for PR Preview to work?

**Review by selfissued — APPROVED** [review]
This is a great start at the simpler JWT-VC 2.0 format that many developers have been requesting.

**Review by mprorock — APPROVED** [review]
big cheer!

**Review by decentralgabe — APPROVED** [review]

**Review by quartzjer — APPROVED** [review]

**Comment by quartzjer** [comment]
This will fix :all-the-things:, woooo!!!!

File: index.html [hunk 1]
```diff
@@ -625,6 +625,54 @@ <h3>Instead of...</h3>
...
+      <section>
+        <h2>Credential</h2>
+        <aside
+          class="example"
+          title="credential (decoded JWT payload)"
+        >
+          <pre data-include="./examples/vc-2.0/credential.json" data-include-format="json"></pre>
+        </aside>
+        <p class="note">
+          The <code>vc</code> and <code>vp</code> claims MUST NOT be present when using the content type <code>credential+json</code>.
```
Comment (mprorock) [inline index.html:657]:
```suggestion
          The <code>Verifiable Credential</code> and <code>Verifiable Presentation</code> claims MUST NOT be present when using the content type <code>credential+json</code>.
```

File: index.html [hunk 2]
```diff
@@ -625,6 +625,54 @@ <h3>Instead of...</h3>
...
+        </p>
+      </section>
+      <section>
+        <h2>Verifiable Credential</h2>
+        <aside
+          class="example"
+          title="verifiable credential"
+        >
+          <p>An encoded verifiable credential (using external proof), decode it with <a href="https://jwt.io/">jwt.io</a>.</p>
```
Comment (mprorock) [inline index.html:666]:
```suggestion
          <p>An encoded Verifiable Credential (using external proof), decode it with <a href="https://jwt.io/">jwt.io</a>.</p>
```

File: index.html (hunk 1 above)
Comment (mprorock) [inline index.html:657]:
```suggestion
          The <code>vc</code> and <code>vp</code> claims <em>MUST NOT</em> be present when using the content type <code>credential+json</code>.
```

**Review by mirceanis — APPROVED** [review]
Very interesting, I like this simpler format.

**Comment by letmaik** [comment]
[5.3.1 Semantic Interoperability](https://www.w3.org/TR/vc-data-model/#semantic-interoperability) says that correctly understanding and parsing a VC in non-JSON-LD clients requires that they first do an exact match on the `@context` array and then continue reading all other properties. [5.4 Data Schemas](https://www.w3.org/TR/vc-data-model/#data-schemas) then goes further and is what I would call a media type, if each credential type had its own media type, that is. On top of that, the [JWT encoding](https://www.w3.org/TR/vc-data-model/#jwt-encoding) allows to re-use some existing JWT claims and later merge them into the decoded credential.

I can see that this is complicated and probably not well implemented in producers and consumers. Part of that may be the optionality of features like credential schemas or the multitude of concepts that have to be looked at to fully validate/understand a VC.

This PR seems to give up on JSON-LD and rather use more top-level JWT claims, either by registering new ones in IANA, or by punting and relying on private claim names. I don't see the concept of schema or context anymore, and the `cty` parameter in the JWT header would not be application-specific either.

I wonder how things like the [Traceability Vocabulary](https://w3c-ccg.github.io/traceability-vocab/#example-2) would work in this world. Or are those not in scope for VC-JWT 2 and a different proof format like [VC-JWS](https://transmute-industries.github.io/vc-jws/) would be reserved for JSON-LD? Will this lead to two parallel VC worlds?

I think having more complex examples would help evaluate this proposal.

File: index.html [hunk 3]
```diff
@@ -625,6 +625,54 @@ <h3>Instead of...</h3>
...
+      <p class="note">
+        This representation relies on claims registered in the
+        <a href="https://www.iana.org/assignments/jwt/jwt.xhtml#claims">IANA JSON Web Token Claims Registry</a> whenever possible.
+      </p>
+      <p class="note">
+        Production of this representation does not use <code>credential+ld+json</code> as an input.
+      </p>
+      <section>
+        <h2>Credential Metadata</h2>
+        <p><a data-cite="rfc7519#section-5.1">typ</a> MUST use the content type <code>verifiable-credential+jwt</code>.</p>
```
Comment (bc-pi) [inline index.html:639]: sorry for bike-shedding but what about "vc+jwt"? 
```suggestion
        <p><a data-cite="rfc7519#section-5.1">typ</a> MUST use the content type <code>vc+jwt</code>.</p>
```

**Comment by OR13** [comment]
@letmaik Thank you for your comments!

> [5.3.1 Semantic Interoperability](https://www.w3.org/TR/vc-data-model/#semantic-interoperability) says that correctly understanding and parsing a VC in non-JSON-LD clients requires that they first do an exact match on the @context array and then continue reading all other properties.

Indeed! 

There would be no "mandatory JSON-LD semantic interoperability" in this format, since there is no mandatory `@context`.

This approach does not secure the "same core data model" that was described in version 1.1, and that was recently given a unique media type to clearly communicate what a "credential" is:  https://github.com/w3c/vc-data-model/pull/1000 .

This proposal secures JSON, not JSON-LD.

This is in comparison to approaches such as:

- https://transmute-industries.github.io/vc-jws/ (not a formal work item yet)
- https://transmute-industries.github.io/vc-cose/ (not a formal work item yet)
- https://w3c.github.io/vc-data-integrity/ (a formal work item on standards track in this working group)

As is noted here: https://github.com/w3c/vc-jwt/pull/44/files#diff-0eb547304658805aad788d320f10bf1f292797b5e6d745a3bf617584da017051R672

If you want to secure `application/credential+ld+json`, this representation is not a good choice for that (at least in the current PR structure).

The only "standards track way" to secure `application/credential+ld+json` currently, is with "data integrity proofs"... I think thats a mistake, I made this repo to explain why:

https://or13.github.io/decanonicalization/

Folks on the mailing list have also suggested that we define a simpler security format for that media type, that does not rely on canonicalization.

- https://lists.w3.org/Archives/Public/public-vc-wg/2023Jan/0030.html
- https://lists.w3.org/Archives/Public/public-vc-wg/2023Jan/0026.html
- https://lists.w3.org/Archives/Public/public-vc-wg/2023Jan/0019.html
- https://lists.w3.org/Archives/Public/public-vc-wg/2023Jan/0012.html

I put all this here to say, lets not distract from what is in the current PR that we are on, as these topics are still being discussed on the mailing lists, and there seems to be strong support for what you are asking for, but enabling that ask, is probably a separate work item (as I note above, its currently not possible with any of the current work items, other than data integrity proofs).

**Comment by TallTed** [comment]
@OR13 — When we quote a snippet, GitHub "helpfully" removes the backticks, so ``` `@context` ``` becomes `@context` within the quote block, and we have to go manually code fence it ... as remains for you to do in your https://github.com/w3c/vc-jwt/pull/44#issuecomment-1413878169

**Review by dwaite — APPROVED** [review]

**Review by danielfett — APPROVED** [review]

**Review by talltree — APPROVED** [review]

File: index.html [hunk 4]
```diff
@@ -625,6 +625,55 @@ <h3>Instead of...</h3>
...
+          The <code>vc</code> and <code>vp</code> claims MUST NOT be present when using the content type <code>credential+json</code>.
+        </p>
+      </section>
+      <section>
+        <h2>Verifiable Credential</h2>
+        <aside
+          class="example"
+          title="verifiable credential"
+        >
+          <p>An encoded verifiable credential (using external proof), decode it with <a href="https://jwt.io/">jwt.io</a>.</p>
```
Comment (bc-pi) [inline index.html:667]: Is it appropriate to link to a vendor created/sponsored site like jwt.io in a w3c doc?

**Comment by SmithSamuelM** [comment]
approve

**Comment by dhh1128** [comment]
LGTM

**Comment by SmithSamuelM** [comment]
@OR13  please correct me if my interpretation of your graph here https://or13.github.io/decanonicalization/ is inaccurate.  An LD signature using URDNA-2015 on an RDF graph derived from JSON-LD where such graph has 81 blank nodes takes 5 mintues to verify. ?

Based on my reading of the literature (see https://marceloarenas.cl/publications/iswc11.pdf) rdf graphs in the wild (depending on the complexity of the information stored in the graph) will have anywhere from near zero percentage to over 85% of the nodes in the graph will be blank nodes. I infer that the recursive (mention gossip) traversal needed for labeling blank nodes in URDNA-2015 can result in an NP Hard labeliing problem when there are large numbers of blank nodes (apparently > 64 is a large number).  

In some applications, no doubt, the json-ld VCs do not create any significant number of blank nodes so verification performance is not an issue. But my concern is that in real world applications like international trade and trade finance the document types and the inter-relationships between documents are both complex which would likely lead to high percentages of blank nodes thereby triggering impractical verification times for a set of documents presented as VCs.  For example, I a real life enterprise example the following documents must be cross validated after they are authenticated and the entities presenting them are verified as authorized to present them: packing list, bill-of-lading, commercial invoice, purchase order, customs declaration, certificate of origin, trade contract, warehouse receipt, transport insurance certificate. Not to mention payment details.  Many enterprises will need to protect their identifiers with multiple-signatures, each a separate proof of authenticity. Enterprises need to designate and authorize representatives that naturally results in chained provenance  of authority for a given representative to sign a given document. This multiplies the number of verifications. So any verification that takes on the order of seconds much less minutes will be impractical at scale.

One could simplify the documents to minimize the number of blank nodes, but isn’t the complexity of links between triples the whole point of using RDF in the first place? In other words the reason one would want to use a semantic reasoning tool like RDF is to automate complex business logic on the set of documents and their interrelationships?

More performant approaches to authentication verification (like JWT proofs on the contents of the VC not the expanded RDF graph) that do not incur the risk of an exponential time cascade for proof verification seem well justified. Hence my approval for this pull request.


As you know I support graph based reasoning models for such purposes, but I do not consider RDF compared to other such methodologies.

**Review by tlodderstedt — APPROVED** [review]
I applaud the direction of this PR as it uses JWS built-in capabilities (like `typ`) to determine the format of the VC in the payload. That makes it easy to use and robust and is natural to developers familiar with JWS and JWT. It is also simple to use this design to envelope binary credentials formats.

It might be worthwhile to evaluate whether a single more fine grained `typ` value would be the better option than using a combination of `typ` and `cty`.

**Review by nadalin — APPROVED** [review]

File: index.html [hunk 5]
```diff
@@ -625,6 +625,55 @@ <h3>Instead of...</h3>
...
+          <pre data-include="./examples/vc-2.0/verifiable-credential.jwt" data-include-format="text"></pre>
+        </aside>
+        <p class="note">
+          Verification of this representation is not required to produce 
+          <code>credential+ld+json</code> or <code>credential-claims-set-1.1+json</code> as an output.
```
Comment (msporny) [inline index.html:672]: Can you point to the definition of "Verification of this representation"? Do you mean JSON Schema-based verification? JWT-based verification? Or something else? It's unclear what the term "verification" here means.

**Review by msporny — CHANGES_REQUESTED** [review]
Minor clarification requested, otherwise, LGTM.

**Review by awoie — APPROVED** [review]
lgtm

**Comment by msporny** [comment]
> @OR13 please correct me if my interpretation of your graph here https://or13.github.io/decanonicalization/ is inaccurate.

Your interpretation is inaccurate, but mostly because 1) the setup for the experiment is flawed, and 2) the way the information is being presented is deeply misleading (the x-axis label doesn't indicate what's actually going on in the code). 

What Orie's test does is it copies n^2 *completely identical objects* into a graph, turns off the safety (that is, it tells the algorithm to use an infinite amount of compute), and then asks the algorithm to create a canonical ordering. 

In other words, it's a completely artificial test -- a poison graph, which we've known about for over a decade (and is why we have safeties in the algorithm). 

To put it in perspective, here's what happens when you have a graph only consisting of blank nodes where each blank node has at least one property with a unique value (that is, this is how it looks when you're dealing with data that's slightly closer to a real world scenario):

https://dlongley.github.io/decanonicalization/

So, for a graph containing 9,801 objects (blank nodes) with zero connectivity between them, your JSON-LD processing + RDF Canonicalization + signing time is 292 milliseconds. Also, who is constructing documents like that in the real world?!

If you have a graph w/ 40 objects (blank nodes) with zero connectivity between them, you're looking at around 11 milliseconds to canonicalize... but again, that's a completely synthetic use case. Show us a real world use case that's dealing with that sort of structure, please.

This also completely ignores the fact that a production deployment has almost certainly got some JSON schema (or roughly equivalent) checking on the incoming object before you even get to full blown JSON-LD / RDF processing... and if you've got someone that's sending data like the above to you -- they're attacking you and you'd employ the same sorts of mitigations you would in a DoS or DDoS scenario. 

Again, security is about layers, and these are some reasonable layers:

1. Rate limit API endpoints with the option to block abusive IP ranges.
2. Perform schema validation (e.g., JSON Schema, message sizes) on incoming data.
3. Then, and only if you have to, perform JSON-LD processing and RDF processing ENSURING that you set an upper limit on compute (which goes for any part of your system... you probably wouldn't accept 100MB JSON document uploads to your API (reasonable limits), or requests to digitally sign 1,000s of objects in a single call (reasonable limits), or public keys that don't fall on an EC curve when doing P-256, or public keys that are susceptible to a twist attack (mandatory attack checking)... the list goes on).

> An LD signature using URDNA-2015 on an RDF graph derived from JSON-LD where such graph has 81 blank nodes takes 5 mintues to verify. ?

Hope the above clarifies why this conclusion is incorrect.

**Comment by OR13** [comment]
@msporny

> Also, who is constructing documents like that in the real world?!

Attackers, like me ; )

> @OR13, it would be helpful if you updated your test with an explanation... 

I added links to your comments here:

https://github.com/OR13/decanonicalization#updates

**Comment by SmithSamuelM** [comment]
@msporny @OR13 Thanks for the clarifications. We need more of these discussions.  @Manu albeit that in your more realistic scenario the exponential rise in verification time is must less pronounced, it is still clearly exponential. This means that there is a ceiling eventually at big enough scale. What that scale is and how close we will ever get to that ceiling is worthy of discussion.  @OR13 speaks to the fundamental security issue of moving verification from verifying the over the wire serialization to a eventual expanded representation is that an attacker can DDOS the expansion process and the expander can't know until post expansion that the attack is unverifiable. So both (exponential scaling ceiling and post expansion security) are concerns that the community should discuss. But these both support my contention that we should have simpler alternatives for verification that do not trigger either of these adverse effects.

**Comment by dlongley** [comment]
@SmithSamuelM,

> This means that there is a ceiling eventually at big enough scale. What that scale is and how close we will ever get to that ceiling is worthy of discussion.

Agreed. Just like there's a ceiling for any other algorithm -- what matters is whether it would be reached for any real use case and whether reaching it for non-use cases (attacks) can be prevented. So far, the W3C RCH WG thinks that real use cases that would hit the ceiling don't exist and non-use cases (attacks) can be prevented.

**Review by tplooker — APPROVED** [review]
LGTM

**Comment by dlongley** [comment]
@OR13,

> Is there an "only documents that are safe to canonize" option I am missing here?
> 
> Or am I supposed to run JSON Schema on the document before calling this method?
> 
> I'd prefer to verify first, then compare to schema.

You should compare to schema first to avoid hitting cryptographic systems and incurring costs there, including potential key resolution costs. A good schema will stop all non-real use cases before they ever get to this layer.

But, as I mentioned to you elsewhere, the RCH WG is currently working on figuring out the common, interoperable way (or if whether we'll just be giving advice) on how to specify poison graph detection. Specifying a way to do it that might be different when that group is finished creates unnecessary churn when schema checks already prevent attacks.

That being said, because there's so much confusion going on here in this thread and there has not been a deep analysis of what's actually going on ... it's created a situation where being responsive and then having to deal with breaking changes is perhaps the way to go. So, in our own libraries, I've exposed the ability to pass in canonize options in the `Ed25519Signature2020` cryptosuite and below is an example of it in use:

```js
    it('should fail to verify a document with a poison graph', async () => {
      const poisonCredential = {...signedCredential};
      const suite = new Ed25519Signature2020({
        canonizeOptions: {maxDeepIterations: 1}
      });
      poisonCredential.alumniOf = poisonData;

      const result = await jsigs.verify(poisonCredential, {
        suite,
        purpose: new AssertionProofPurpose(),
        documentLoader
      });
      expect(result.verified).to.be.false;
      const {error} = result.results[0];
      expect(error.message).to.include('Maximum deep iterations');
    });
```

**Comment by OR13** [comment]
I suggest we move the "data poisoning conversation" to  https://github.com/w3c/vc-data-integrity

This repo does not require any URDNA2015 processing.

**Comment by SmithSamuelM** [comment]
@dlongley 

> But, at the end of the day there's a ceiling for any algorithm --

It don't think that is a fair conclusion. Clearly exponential scaling algorithms as a class have much lower ceilings than linear or sub-linear scaling algorithms. And for the most part linear and sublinear scaling algorithms usually, never reach a ceiling because the computation capability grows faster than the application demand. Whereas exponentially scaling algorithms can reach their practical ceilings very quickly relative to practical growth in computational capability, so we have to do due diligence when considering exponential scaling algorithms.  Lumping all algorithms classes together is not helpful here.  Signatures on over-the-wire serializations such as a JWT proofs are linear with the length of the serialization whereas URDNA-2015 is exponential with the complexity of the RDF graph structure. These are not comparable concerns. The former I don't have to worry about the latter I do.

**Comment by dlongley** [comment]
@SmithSamuelM, if you'd like to keep discussing, please take it over to the issue Orie linked to.

File: examples/vc-2.0/credential-metadata.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
+  "kid": "urn:example:issuer#key-0",
+  "alg": "ES256",
+  "typ": "verifiable-credential+jwt",
+  "cty": "credential+json"
```
Comment (dlongley) [inline examples/vc-2.0/credential-metadata.json:5]: This doesn't use the right `cty` name that was merged elsewhere.

File: index.html [hunk 7]
```diff
@@ -625,6 +625,51 @@ <h3>Instead of...</h3>
...
+        <a href="https://www.iana.org/assignments/jwt/jwt.xhtml#claims">IANA JSON Web Token Claims Registry</a> whenever possible.
+      </p>
+      <p class="note">
+        Production of this representation does not use <code>credential+ld+json</code> as an input.
+      </p>
+      <section>
+        <h2>Credential Metadata</h2>
+        <p><a data-cite="rfc7519#section-5.1">typ</a> MUST use the content type <code>verifiable-credential+jwt</code>.</p>
+        <p class="issue" data-number="45"></p>
+        <p>If present, <a data-cite="rfc7519#section-5.2">cty</a> SHOULD use the content type <code>credential+json</code>.</p>
```
Comment (dlongley) [inline index.html:641]: This doesn't use the right content type name that was merged elsewhere.

File: index.html (hunk 1 above)
Comment (dlongley) [reply index.html:658]: This doesn't use the right content type name that was merged elsewhere.

File: index.html (hunk 7 above)
Comment (msporny) [inline index.html:641]:
```suggestion
        <p>If present, <a data-cite="rfc7519#section-5.2">cty</a> SHOULD use the content type <code>credential-claims-set+json</code>.</p>
```

File: index.html [hunk 8]
```diff
@@ -625,6 +625,51 @@ <h3>Instead of...</h3>
...
+      <section>
+        <h2>Credential</h2>
+        <aside
+          class="example"
+          title="credential (decoded JWT payload)"
+        >
+          <pre data-include="./examples/vc-2.0/credential.json" data-include-format="json"></pre>
+        </aside>
+        <p class="note">
+          The <code>vc</code> and <code>vp</code> claims MUST NOT be present when using the content type <code>credential+json</code>.
```
Comment (msporny) [inline index.html:658]:
```suggestion
          The <code>vc</code> and <code>vp</code> claims MUST NOT be present when using the content type <code>credential-claims-set+json</code>.
```

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-02-08-vcwg#section3-7) on 2023-02-08

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 3.7. Add version 2 (pr vc-jwt#44)

_See github pull request [vc-jwt#44](https://github.com/w3c/vc-jwt/pull/44)._




**Orie Steele:** One open p.r. for vc jwt.  
… build on previous discussions around media type.  
… some off topic conversations moved to other issues.  
… expect to spend time at f2f discussing this.  

**Brent Zundel:** any other updates about work items.  

</details>

**Comment by OR13** [comment]
@msporny I have merged your suggestions, can you please re-review?

**Review by dlongley — APPROVED** [review]
Thanks!

I will note here that I still think using `cty` here doesn't follow best practice / existing RFCs for JWTs, especially given that `typ` is present and *does* follow best practice for "[Explicit Typing](https://www.rfc-editor.org/rfc/rfc8725#name-use-explicit-typing)", but I'm not blocking this PR.

**Review by msporny — APPROVED** [review]
Thanks for the updates, I note the same concerns as @dlongley has spoken to on the calls and in this PR... the current IETF JWT RFCs advise against how `cty` is being used here. It sounds like vc-jwt is going to ignore that guidance, so you might want to say why in the spec... in a separate PR (new issue please?). 

Otherwise, LGTM.

**Comment by TallTed** [comment]
@selfissued — I am surprised to see that this PR was merged without the [new issue requested by](https://github.com/w3c/vc-jwt/pull/44#pullrequestreview-1291555730) @msporny having first been created. Please do create that new issue now.

**Comment by OR13** [comment]
@TallTed filed: https://github.com/w3c/vc-jwt/issues/47

**Pull request merged**